Routes should be able to specify which port they desire

smarterclayton opened the pull request:
Filter UDP ports from template router by default.

WIP needs F5 support and still need new-app

@ramr @pweil-

smarterclayton commented:
Fixes #4603

pweil- commented on plugins/router/template/router.go:
```diff
@@ -144,6 +145,20 @@ func newTemplateRouter(cfg templateRouterCfg) (*templateRouter, error) {
+	endpoints := make([]Endpoint, 0, len(svc.EndpointTable))
+	for i := range svc.EndpointTable {
+		endpoint := svc.EndpointTable[i]
+		if endpoint.PortName == alias.PreferPort || endpoint.Port == alias.PreferPort {
```
I think I'm missing something here.  How is the use case of specifying multiple ports in my route api object supported by our template router?  What I see is:

1.  The endpoints are created from all endpoints in the endpoints object (one more "endpoints" for good measure) but filtering anything that is UDP if configured to do so
2.  A preferred port is always set for specs with ports > 0
3.  The template in our default impl will call this method which will only include the preferred port which is either a single name or port number

I'm not seeing how my Route.Spec.Ports settings filter down to the template to select a subset of the endpoint table.  Perhaps this is the IP part of the WIP.

smarterclayton replied:
The preferred port is what causes the filter to happen.  You either get all
endpoints (no ports) or endpoints that match the first port you define.

On Thu, Sep 10, 2015 at 1:12 PM, Paul Weil <notifications@github.com> wrote:

> In plugins/router/template/router.go
> <https://github.com/openshift/origin/pull/4606#discussion_r39185947>:
>
> > @@ -144,6 +145,20 @@ func newTemplateRouter(cfg templateRouterCfg) (*templateRouter, error) {
> >  	return router, nil
> >  }
> >
> > +func endpointsForAlias(alias ServiceAliasConfig, svc ServiceUnit) []Endpoint {
> > +	if len(alias.PreferPort) == 0 {
> > +		return svc.EndpointTable
> > +	}
> > +	endpoints := make([]Endpoint, 0, len(svc.EndpointTable))
> > +	for i := range svc.EndpointTable {
> > +		endpoint := svc.EndpointTable[i]
> > +		if endpoint.PortName == alias.PreferPort || endpoint.Port == alias.PreferPort {
>
> I think I'm missing something here. How is the use case of specifying
> multiple ports in my route api object supported by our template router?
> What I see is:
>
>    1. The endpoints are created from all endpoints in the endpoints
>    object (one more "endpoints" for good measure) but filtering anything that
>    is UDP if configured to do so
>    2. A preferred port is always set for specs with ports > 0
>    3. The template in our default impl will call this method which will
>    only include the preferred port which is either a single name or port number
>
> I'm not seeing how my Route.Spec.Ports settings filter down to the
> template to select a subset of the endpoint table. Perhaps this is the IP
> part of the WIP.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4606/files#r39185947>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

pweil- replied:
Right, it filters to the preferred port.  So if I define multiple ports in my route it doesn't do anything with the 0+n ports.  Just a bit confusing that I can configure for multiple ports and it ignores them.  It doesn't seem to be a reasonable use case to want 80 and 8888 of your pod on the same route anyway but if we did want to support it we'd need to plumb through the route.Ports.   Just another reason to get rid of this internal representation.

smarterclayton replied:
On Sep 10, 2015, at 5:23 PM, Paul Weil <notifications@github.com> wrote:

In plugins/router/template/router.go
<https://github.com/openshift/origin/pull/4606#discussion_r39215815>:

> @@ -144,6 +145,20 @@ func newTemplateRouter(cfg templateRouterCfg) (*templateRouter, error) {
>  	return router, nil
>  }
>
> +func endpointsForAlias(alias ServiceAliasConfig, svc ServiceUnit) []Endpoint {
> +	if len(alias.PreferPort) == 0 {
> +		return svc.EndpointTable
> +	}
> +	endpoints := make([]Endpoint, 0, len(svc.EndpointTable))
> +	for i := range svc.EndpointTable {
> +		endpoint := svc.EndpointTable[i]
> +		if endpoint.PortName == alias.PreferPort || endpoint.Port == alias.PreferPort {

Right, it filters to the preferred port. So if I define multiple ports in
my route it doesn't do anything with the 0+n ports. Just a bit confusing
that I can configure for multiple ports and it ignores them. It doesn't
seem to be a reasonable use case to want 80 and 8888 of your pod on the
same route anyway but if we did want to support it we'd need to plumb
through the route.Ports. Just another reason to get rid of this internal
representation.



Ideally, a sophisticated router would match route ports to features.
However, it's not worth the complexity.  I left multiple ports because we
will have to deal with it in the future (f5)


—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4606/files#r39215815>.

ramr replied:
Catching up this week. Should be easy to add multiple ports in here, no? 
Let the ```ServiceAliasConfig``` have a list of preferred ports and 
add 'em to a map[string]bool ```mapR``` here and then later just change this check: 
```endpoint.PortName == alias.PreferPort || endpoint.Port == alias.PreferPort``` 
to ```mapR[endpoint.PortName]  || mapR[endpoint.Port]``` or something to that effect.

smarterclayton replied:
I think the concern is that the different ports mean something different.
Why would I have both 80 and 443 exposed on a service?  I might have 8080
and 8081 that I want round robin-ed, but I can't think of a real reason to
do that.  There was no scenario I could construct for a route where we
wanted to expose 1 port to multiple backing endpoint ports.  Instead, all
the scenarios were where we wanted to have 80+443, then maybe expose
2800-2900 for additional extension.  But even then, it's not clear that
extra port is that useful.

On Tue, Sep 15, 2015 at 2:34 PM, Ram Ranganathan <notifications@github.com>
wrote:

> In plugins/router/template/router.go
> <https://github.com/openshift/origin/pull/4606#discussion_r39547794>:
>
> > @@ -144,6 +145,20 @@ func newTemplateRouter(cfg templateRouterCfg) (*templateRouter, error) {
> >  	return router, nil
> >  }
> >
> > +func endpointsForAlias(alias ServiceAliasConfig, svc ServiceUnit) []Endpoint {
> > +	if len(alias.PreferPort) == 0 {
> > +		return svc.EndpointTable
> > +	}
> > +	endpoints := make([]Endpoint, 0, len(svc.EndpointTable))
> > +	for i := range svc.EndpointTable {
> > +		endpoint := svc.EndpointTable[i]
> > +		if endpoint.PortName == alias.PreferPort || endpoint.Port == alias.PreferPort {
>
> Catching up this week. Should be easy to add multiple ports in here, no?
> Let the ServiceAliasConfig have a list of preferred ports and
> add 'em to a map[string]bool mapR here and then later just change this
> check:
> endpoint.PortName == alias.PreferPort || endpoint.Port == alias.PreferPort
> to mapR[endpoint.PortName] || mapR[endpoint.Port] or something to that
> effect.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4606/files#r39547794>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton commented:
Refactored to just have a single optional port - this makes us look a lot more like Ingress (which is going to be 1..N routes).  I added test cases and validation, this should be ready for review.  A follow up issue will come in and add route defaulting to all creation paths such that the route defaults to at least one of the ports (instead of showing all ports).

smarterclayton commented:
[test]

smarterclayton commented:
@ramr @pweil- you are making me sad panda on review

pweil- commented:
This actually makes way more sense to me with the single optional port.  I say :+1:

ramr commented:
LGTM

smarterclayton commented:
[merge]

On Oct 1, 2015, at 10:31 PM, Ram Ranganathan <notifications@github.com>
wrote:

LGTM

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4606#issuecomment-144839070>.
